src/component/isolation.rs

```diff
@@ -93,46 +97,133 @@ impl ProcessIsolationController for PythonUnsafeController {
     fn boot_process(&self) -> Result<Box<dyn IsolatedProcessHandle>, WorkerError> {
         let pipe = self.pipe_creator.new_pipe()?;
 
-        let c_in = canonicalize(pipe.component_input_file())?.into_os_string();
-        let c_out = canonicalize(pipe.component_output_file())?.into_os_string();
+        let c_in = canonicalize(pipe.component_input_file())?
+            .into_os_string()
+            .into_string()
+            .map_err(WorkerErrorKind::OsStringConversion)?;
+        let c_out = canonicalize(pipe.component_output_file())?
+            .into_os_string()
+            .into_string()
+            .map_err(WorkerErrorKind::OsStringConversion)?;
 
         let subprocess = Popen::create(
-            &[
-                "python3",
-                "-u",
-                &self.executable_file,
-                &c_in.to_string_lossy(),
-                &c_out.to_string_lossy(),
-            ],
+            &["python3", "-u", &self.executable_file, &c_in, &c_out],
             PopenConfig::default(),
         )?;
 
-        Ok(Box::new(PythonUnsafeHandle { subprocess, pipe }))
+        Ok(Box::new(PipedProcessHandle { subprocess, pipe }))
     }
 }
 
 #[derive(Debug)]
-pub struct PythonUnsafeHandle {
+struct DockerArchiveController {
+    pipe_creator: NamedPipeCreator,
+    docker_image_tag: String,
+}
+
+impl DockerArchiveController {
+    pub fn new(docker_tar_file_path: &str) -> Result<Self, WorkerError> {
+        if cfg!(target_os = "macos") {
+            warn!("using docker for isolation on macOS likely will not work!!!");
+        }
+
+        let load_result = Exec::cmd("docker")
+            .args(&["load", "-q", "--input", docker_tar_file_path])
+            .stdout(Redirection::Pipe)
+            .capture()?;
+        let load_stdout = String::from_utf8(load_result.stdout)?;
+        let load_stderr = String::from_utf8(load_result.stderr)?;
+
+        if !load_result.exit_status.success() {
+            return Err(
+                WorkerErrorKind::Docker(load_result.exit_status, load_stdout, load_stderr).into(),
+            );
+        }
+
+        // stdout looks like
+        // Loaded image: tag\n
+        let trimmed_stdout = load_stdout.trim();
+        let split_idx = "Loaded image: ".len();
+        if split_idx >= trimmed_stdout.len() {
+            return Err(
+                WorkerErrorKind::Docker(load_result.exit_status, load_stdout, load_stderr).into(),
+            );
+        }
+        let tag = &trimmed_stdout.trim()[split_idx..];
```

Comment:
If someone can figure out how to regex this well, I'd love them forever.

Reply:
If the tag has a fixed length this could be a little smaller by getting substring((stringlen-2-taglen),(stringlen-2))

Reply:
I found a regexy solution that I'll push along with other PR improvements